flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -468,6 +471,69 @@ void sinkLoopIVArgs(mlir::ConversionPatternRewriter &rewriter,
     ++idx;
   }
 }
+
+/// Collects values that are local to a loop: "loop-local values". A loop-local
+/// value is one that is used explicitely inside the loop but allocated outside
+/// of it. This usually corresponds to temporary values that are used inside the
+/// loop body for initialzing other variables for example.
+///
+/// \param [in] doLoop - the loop within which the function searches for locally
+/// destroyed values.
+///
+/// \param [out] locals - the list of loop-local values detected for \p doLoop.
+void collectLoopLocalValues(fir::DoLoopOp doLoop,
+                            llvm::SetVector<mlir::Value> &locals) {
+  doLoop.walk([&](mlir::Operation *op) {
+    for (mlir::Value operand : op->getOperands()) {
+      bool isLocal = true;
+
+      if (!mlir::isa_and_present<fir::AllocaOp>(operand.getDefiningOp()))
+        continue;
+
+      // Values defined inside the loop are not interesting since they do not
+      // need to be localized.
+      if (doLoop->isAncestor(operand.getDefiningOp()))
+        continue;
+
+      for (auto *user : operand.getUsers()) {
+        if (!doLoop->isAncestor(user)) {
+          isLocal = false;
+          break;
+        }
+      }
+
+      if (isLocal)
+        locals.insert(operand);
+    }
+  });
+}
+
+/// For a "loop-local" value \p local within a loop's scope, localizes that
+/// value within the scope of the parallel region the loop maps to. Towards that
+/// end, this function allocates a private copy of \p local within \p
+/// allocRegion.
+///
+/// \param local - the locally destroyed value within a loop's scope (see
+/// collectLocallyDestroyedValuesInLoop).
+///
+/// \param allocRegion - the parallel region where \p local's allocation will be
+/// cloned (i.e. privatized).
+///
+/// \param rewriter - builder used for updating \p allocRegion.
+///
+/// \param mapper - mapper to track updated references \p local within \p
+/// allocRegion.
+void localizeLoopLocalValue(mlir::Value local, mlir::Region &allocRegion,
+                            mlir::ConversionPatternRewriter &rewriter,
+                            mlir::IRMapping &mapper) {
+  mlir::IRRewriter::InsertPoint ip = rewriter.saveInsertionPoint();
+  rewriter.setInsertionPointToStart(&allocRegion.front());
+  mlir::Operation *newLocalDef = rewriter.clone(*local.getDefiningOp(), mapper);
```

Comment:
We should be able to use `rewriter.moveOpAfter()` and avoid making copies of the alloca, since we checked that all uses of the allocation are inside of that region. Then comments should be updated to say that allocas are sunk or moved into the parallel region rather than cloned in it.